this change moves stack of compression methods, now global variable

Sashan opened the pull request:
in libssl into crypto context. This change is rquired by atexit PR.

mattcaswell commented on ssl/ssl_init.c:
```diff
@@ -78,7 +78,6 @@ static void ssl_library_stop(void)
 #ifndef OPENSSL_NO_COMP
         OSSL_TRACE(INIT, "ssl_library_stop: "
                    "ssl_comp_free_compression_methods_int()\n");
-        ssl_comp_free_compression_methods_int();
```
This whole function can be removed since it is no longer doing anything

mattcaswell reviewed (changes requested):
When I build this I get linker errors:

````
/usr/bin/ld: ./libssl.so: undefined reference to `ossl_lib_ctx_get_data'
/usr/bin/ld: ./libssl.so: undefined reference to `ossl_find_compression'
/usr/bin/ld: ./libssl.so: undefined reference to `ossl_lib_ctx_set0_compression_methods'
````

This is because these symbols are not exported from libcrypto. It seems to me the only symbol that you actually need to export from libcrypto is  `ossl_lib_ctx_get_data'. The other two could probably just be in libssl.

We have as yet not had the concept of internal symbols that are exported from libcrypto and are privately used by libssl. All symbols exported from libcrypto are public and form part of the API. With that in mind I would actually create a new public libcrypto function `OSSL_LIB_CTX_get_data` as a simple wrapper around `ossl_lib_ctx_get_data`. The public documentation would say that, although this is a public function, all the indexes that you can pass are reserved for internal use and so applications should not call this function. We might at some point in the future extend it with the idea to support application supplied indexes - but we don't have to worry about that now.

Sashan commented:
> When I build this I get linker errors:
> 
> ```
> /usr/bin/ld: ./libssl.so: undefined reference to `ossl_lib_ctx_get_data'
> /usr/bin/ld: ./libssl.so: undefined reference to `ossl_find_compression'
> /usr/bin/ld: ./libssl.so: undefined reference to `ossl_lib_ctx_set0_compression_methods'
> ```
>
I've not noticed those errors on OpenBSD/clang tool chain. I can see those errors on Debian-12/gcc tool chain. I agree we can remove `ossl_find_compression()` which implementation can be kept in libssl. I will also need public version of `ossl_lib_ctx_set0_compression_methods()` because it is a workhorse for  `SSL_COMP_set0_compression_methods()` function. Unless we want to remove `SSL_COMP_set0_compression_methods()` from libssl API.

mattcaswell commented:
> I will also need public version of ossl_lib_ctx_set0_compression_methods() because it is a workhorse for SSL_COMP_set0_compression_methods() function. Unless we want to remove SSL_COMP_set0_compression_methods() from libssl API.

I'm not sure we need to move this to libcrypto.

What if the libssl data we stored in the libctx was actually a `STACK_OF(OSSL_COMP) **`? Then `SSL_COMP_set0_compression_methods()` could become:

```` C
STACK_OF(SSL_COMP) *SSL_COMP_set0_compression_methods(STACK_OF(SSL_COMP)
                                                      *meths)
{
    STACK_OF(OSSL_COMP) **data = OSSL_LIB_CTX_get_data(NULL, OSSL_LIB_CTX_COMP_METHODS);
    STACK_OF(SSL_COMP) *old_meths = (STACK_OF(SSL_COMP) *)*data;

    *data = (STACK_OF(OSSL_COMP) *)meths;
    return old_meths;
}
````

Sashan commented:
The updated PR addresses all suggestions from @mattcaswell 

I failed to figure out how to hook newly added file `crypto/comp_methods.c` with FIPS module build. Therefore the PR currently disables compression methods for FIPS. With out this workaround in place the fips module fails to load due to unresolved symbols (ossl_load_builtin_compressions()). Perhaps we should dissolve `crypto/comp_methods.c` into existing source code files in libcrypto.

Sashan commented:
> > I will also need public version of ossl_lib_ctx_set0_compression_methods() because it is a workhorse for SSL_COMP_set0_compression_methods() function. Unless we want to remove SSL_COMP_set0_compression_methods() from libssl API.
> 
> I'm not sure we need to move this to libcrypto.
> 
> What if the libssl data we stored in the libctx was actually a `STACK_OF(OSSL_COMP) **`? Then `SSL_COMP_set0_compression_methods()` could become:
> 
> ```c
> STACK_OF(SSL_COMP) *SSL_COMP_set0_compression_methods(STACK_OF(SSL_COMP)
>                                                       *meths)
> {
>     STACK_OF(OSSL_COMP) **data = OSSL_LIB_CTX_get_data(NULL, OSSL_LIB_CTX_COMP_METHODS);
>     STACK_OF(SSL_COMP) *old_meths = (STACK_OF(SSL_COMP) *)*data;
> 
>     *data = (STACK_OF(OSSL_COMP) *)meths;
>     return old_meths;
> }
> ```

I'll update PR shortly to see how it will look. On the other hand I'm not sure if those changes are heading in desired direction. I feel it goes against the concept 'all data are private' to OSSL_LIB_CTX. Changing `OSSL_LIB_CTX_get_data()` so it returns address of `OSSL_LIB_CTX` member makes auditing/reasoning about the code harder. Explicit _set_ function is better in my opinion, because it makes reasoning/auditing the code bit easier.

mattcaswell commented:
> I feel it goes against the concept 'all data are private' to OSSL_LIB_CTX.

That isn't actually the current design of OSSL_LIB_CTX. Currently the context.c code mostly has a bunch of internal members which are "black boxes" as far as context.c is concerned. The internals of the black box are private to the module that uses it. E.g. consider `ctx->drbg`. The context.c code knows nothing about what that data member is - rather this is managed by the crypto/rand/rand_lib.c code where we discover that `ctx->drbg` is actually a `RAND_GLOBAL` structure.

mattcaswell commented on crypto/comp_methods.c:
```diff
@@ -0,0 +1,62 @@
+#define OSSL_COMP_ZLIB_IDX       1
+#define OSSL_COMP_NUM_IDX        2
+
+static int sk_comp_cmp(const OSSL_COMP *const *a, const OSSL_COMP *const *b)
```
This whole function needs to be inside `#ifdef OPENSSL_NO_COMP` - it is causing the one remaining CI failure.

mattcaswell commented on crypto/comp_methods.c:
```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2025 The OpenSSL Project Authors. All Rights Reserved.
```
Getting a bit ahead of ourselves here aren't we?

mattcaswell commented on crypto/comp_methods.c:
```diff
@@ -0,0 +1,62 @@
+            comp->name = COMP_get_name(method);
+            if (!sk_OSSL_COMP_push(comp_methods, comp))
+                OPENSSL_free(comp);
+            sk_OSSL_COMP_sort(comp_methods);
```
This seems redundant since we've only added one method!

mattcaswell commented on crypto/comp_methods.c:
```diff
@@ -0,0 +1,62 @@
+    return comp_methods;
+}
+
+STACK_OF(OSSL_COMP) *ossl_comp_get_compression_methods(OSSL_LIB_CTX *ctx)
```
This comment is outstanding

mattcaswell commented on doc/man3/OSSL_LIB_CTX.pod:
```diff
@@ -111,6 +113,11 @@ in the mean time.  This means that the calling thread must not free the
 library context that was the default at the start of the async job before
 that job has finished.
 
+OSSL_LIB_CTX_get_data() returns address represeing desired item kept in
```
returns **an** address **representing** **the** desired item kept in **the** library context.

mattcaswell commented on doc/man3/OSSL_LIB_CTX.pod:
```diff
@@ -111,6 +113,11 @@ in the mean time.  This means that the calling thread must not free the
 that job has finished.
 
+OSSL_LIB_CTX_get_data() returns address represeing desired item kept in
+library context.  The index argument refers to context member which is
```
refers to **a** context member

mattcaswell commented on doc/man3/OSSL_LIB_CTX.pod:
```diff
@@ -111,6 +115,15 @@ in the mean time.  This means that the calling thread must not free the
+OSSL_LIB_CTX_get_data() returns pointer to data/object member bound to context
+instance (ctx). The index argument refers to context member which is to
+be retrieved. The values for index are private to OpenSSL currently.
+If ctx is NULL then function operates on default crypto library context.
```
This comment is outstanding

mattcaswell commented on doc/man3/OSSL_LIB_CTX.pod:
```diff
@@ -111,6 +113,11 @@ in the mean time.  This means that the calling thread must not free the
 
+OSSL_LIB_CTX_get_data() returns address represeing desired item kept in
+library context.  The index argument refers to context member which is
+to be retrieved. The values for index are private to OpenSSL currently.
```
```suggestion
to be retrieved. The values for index are all private to OpenSSL currently and so
applications should not typically call this function.
```

mattcaswell commented on doc/man3/OSSL_LIB_CTX.pod:
```diff
@@ -121,10 +128,16 @@ OSSL_LIB_CTX_free() doesn't return any value.
 
 OSSL_LIB_CTX_load_config() returns 1 on success, 0 on error.
 
+OSSL_LIB_CTX_get_data() returns a memory address which interpretation
+depends on caller.
```
```suggestion
OSSL_LIB_CTX_get_data() returns a memory address whose interpretation
depends on the index.
```

mattcaswell replied:
This comment is still outstanding

mattcaswell commented on doc/man3/OSSL_LIB_CTX.pod:
```diff
@@ -121,10 +128,16 @@ OSSL_LIB_CTX_free() doesn't return any value.
 
 All of the functions described on this page were added in OpenSSL 3.0.
 
+OSSL_LIB_CTX_get_data() and OSSL_LIB_CTX_set0_compression_methods()
```
`OSSL_LIB_CTX_set0_compression_methods` is no longer exposed.

mattcaswell commented on include/internal/stack_of_comp_methods.h.in:
```diff
@@ -0,0 +1,21 @@
+#ifndef OPENSSL_STACK_OF_COMP_METHODS_H
+
+{-
+    generate_stack_macros("OSSL_COMP");
```
IIRC I don't think this is necessary for internal stacks. I think you can just do `DEFINE_STACK_OF(OSSL_COMP)` directly in cryptlib.h and remove this header entirely.

Sashan commented on crypto/context.c:
```diff
@@ -48,6 +51,7 @@ struct ossl_lib_ctx_st {
     void *thread_event_handler;
     void *fips_prov;
 #endif
+    void *comp_methods;
```
I opted for void as it is easier to handle with respect to typecast (OSSL_COMP vs. SSL_COMP)

Sashan commented on doc/man3/OSSL_LIB_CTX.pod:
```diff
@@ -111,6 +115,15 @@ in the mean time.  This means that the calling thread must not free the
+be retrieved. The values for index are private to OpenSSL currently.
+If ctx is NULL then function operates on default crypto library context.
+
+OSSL_LIB_CTX_set0_compression_methods() function swaps compression methods
```
function `OSSL_LIB_CTX_set0_compression_methods()` got removed `

Sashan commented on doc/man3/OSSL_LIB_CTX.pod:
```diff
@@ -121,10 +134,19 @@ OSSL_LIB_CTX_free() doesn't return any value.
 OSSL_LIB_CTX_load_config() returns 1 on success, 0 on error.
 
+OSSL_LIB_CTX_get_data() returns pointer to desired object associated with
+library context instance.
```
the caller of `OSSL_LIB_CTX_get_data() knows how to interpret the pointer. the pointer can be either pointer to object bound to context or pointer-to-pointer (address in library context) where pointer to desired object is kept.

mattcaswell commented on include/internal/cryptlib.h:
```diff
@@ -161,4 +171,8 @@ char *ossl_buf2hexstr_sep(const unsigned char *buf, long buflen, char sep);
                                    const char sep);
 
+STACK_OF(OSSL_COMP) *ossl_load_builtin_compressions(void);
+STACK_OF(OSSL_COMP) *ossl_comp_get_compression_methods(OSSL_LIB_CTX *ctx);
```
This line can be removed now.

mattcaswell reviewed (changes requested):
Looks good. Just a few minor nits remain.

t8m commented on include/openssl/ssl.h.in:
```diff
@@ -236,7 +236,7 @@ typedef struct ssl_cipher_st SSL_CIPHER;
 typedef struct tls_sigalgs_st TLS_SIGALGS;
 typedef struct ssl_conf_ctx_st SSL_CONF_CTX;
-typedef struct ssl_comp_st SSL_COMP;
+typedef struct ossl_comp_st SSL_COMP;
```
I would not change the typedef from ssl_comp_st to ossl_comp_st. Just move it to openssl/comp.h and keep using SSL_COMP everywhere instead of OSSL_COMP.

t8m replied:
Those casts from OSSL_COMP to SSL_COMP are awkward and unnecessary if we just do not introduce the new type.

t8m commented on include/openssl/comp.h:
```diff
@@ -56,6 +56,16 @@ const BIO_METHOD *BIO_f_brotli(void);
 const BIO_METHOD *BIO_f_zstd(void);
 # endif
 
+typedef struct ssl_comp_st {
+    int id;
+    const char *name;
+    COMP_METHOD *method;
+} SSL_COMP;
```
The structure content should be kept internal -> into `internal/comp.h`

t8m commented on include/openssl/comp.h:
```diff
@@ -56,6 +56,16 @@ const BIO_METHOD *BIO_f_brotli(void);
+
+STACK_OF(SSL_COMP);
+
+DEFINE_STACK_OF(SSL_COMP)
```
This must be generated through the `generate_stack_macros("SSL_COMP")` perl generator by moving that from ssl.h.in -> i.e., it requires generating comp.h from comp.h.in

mattcaswell replied:
I'm wondering why we are bothering to move it? Why not just keep it in ssl.h.in?

t8m replied:
You need to be able to release the stack from libcrypto, so that means the definition needs to be in a header that code linked into libcrypto can include. IMO ssl.h should not be included.

mattcaswell replied:
> IMO ssl.h should not be included.

Why not if we're only using it for some typedefs etc? Moving things to comp.h is problematic because it is excluded in the case of a no-comp build - whereas `SSL_COMP` should not be.

Sashan replied:
the updated PR includes all suggestions from t8m. the code in `comp.h.in` got reshuffled so only relevant portion in that file is protected by OPENSSL_NO_COMP guard. I did test and build it:
    - perl ./Configure --strict-warnings enable-zlib enable-comp 
    - perl ./Configure --strict-warnings no-zlib no-brotli no-zstd no-comp 
 both versions did pass build and `make test`

t8m commented on ssl/ssl_ciph.c:
```diff
@@ -1988,17 +1939,19 @@ uint16_t SSL_CIPHER_get_protocol_id(const SSL_CIPHER *c)
-    }
-    return NULL;
+    srch_key.id = n;
+    i = sk_SSL_COMP_find((STACK_OF(SSL_COMP) *)sk, &srch_key);
```
Is the cast still necessary?

t8m commented on ssl/ssl_ciph.c:
```diff
@@ -1988,17 +1939,19 @@ uint16_t SSL_CIPHER_get_protocol_id(const SSL_CIPHER *c)
+    srch_key.id = n;
+    i = sk_SSL_COMP_find((STACK_OF(SSL_COMP) *)sk, &srch_key);
+    if (i >= 0)
+        ctmp = (SSL_COMP *)sk_SSL_COMP_value((STACK_OF(SSL_COMP) *)sk, i);
```
Are the casts necessary? They should not be.

t8m reviewed (changes requested):
Just a few final tweaks needed

t8m commented on include/openssl/ssl.h.in:
```diff
@@ -976,7 +974,6 @@ extern "C" {
  */
 {-
     generate_const_stack_macros("SSL_CIPHER")
```
isn't this missing `;` ?

t8m commented on include/openssl/comp.h.in:
```diff
@@ -1,12 +1,20 @@
  */
 
+{-
+use OpenSSL::stackhash qw(generate_stack_macros generate_const_stack_macros);
```
generate_const_stack_macros in't needed

t8m commented on include/openssl/comp.h.in:
```diff
@@ -1,12 +1,20 @@
 /*
- * Copyright 2015-2018 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Copyright 4 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2002, Oracle and/or its affiliates. All rights reserved
+ * Copyright 2005 Nokia. All rights reserved.
```
I wouldn't add this. None of the things moved were IMO added by these copyright holders.

t8m commented on include/openssl/comp.h.in:
```diff
@@ -1,12 +1,20 @@
 /*
- * Copyright 2015-2018 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
```
The original copyright should be kept `2015-2024`

t8m commented on ssl/ssl_ciph.c:
```diff
@@ -2021,34 +1976,44 @@ int SSL_COMP_add_compression_method(int id, COMP_METHOD *cm)
-    return ssl_comp_methods;
+    STACK_OF(SSL_COMP) **rv;
+
+    rv = (STACK_OF(SSL_COMP) **) OSSL_LIB_CTX_get_data(NULL,
```
Nit: no space after `**)`

t8m reviewed (approved):
@mattcaswell IMO this is OK. Maybe the SSL_COMP stack declarations could have been handled differently but I see this as cleaner way.

mattcaswell commented on include/openssl/comp.h.in:
```diff
@@ -46,19 +51,26 @@ COMP_METHOD *COMP_brotli_oneshot(void);
+typedef struct ssl_comp_st SSL_COMP;
+
+{-
+    generate_stack_macros("SSL_COMP");
```
I would have preferred to keep this in ssl.h.in. But ok.

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Merged to the master branch with reworded commit message. Thank you.
